rspace/src/main/scala/coop/rchain/rspace/RSpace.scala

```diff
@@ -60,84 +60,120 @@ class RSpace[F[_], C, P, E, A, R, K] private[rspace] (
-                c -> Random.shuffle(store.getData(txn, Seq(c)).zipWithIndex)
-              }.toMap
-            }
-
-            val options: Either[E, Option[Seq[DataCandidate[C, R]]]] =
-              extractDataCandidates(channels.zip(patterns), channelToIndexedData, Nil).sequence
-                .map(_.sequence)
-
-            options match {
-              case Left(e) =>
-                Left(e)
-              case Right(None) =>
-                store
-                  .withTxn(store.createTxnWrite()) { txn =>
-                    store.putWaitingContinuation(
-                      txn,
-                      channels,
-                      WaitingContinuation(patterns, continuation, persist, consumeRef)
-                    )
-                    for (channel <- channels)
-                      store.addJoin(txn, channel, channels)
-                  }
-                logger.debug(s"""|consume: no data found,
+                       options = extractDataCandidates(
+                         channels.zip(patterns),
+                         channelToIndexedData,
+                         Nil
+                       ).sequence
+                         .map(_.sequence)
+                       result <- options match {
+                                  case Left(e) =>
+                                    Left(e).pure[F]
+                                  case Right(None) =>
+                                    store
+                                      .withTxnF(store.createTxnWriteF()) { txn =>
+                                        store.putWaitingContinuation(
+                                          txn,
+                                          channels,
+                                          WaitingContinuation(
+                                            patterns,
+                                            continuation,
+                                            persist,
+                                            consumeRef
+                                          )
+                                        )
+                                        for (channel <- channels)
+                                          store.addJoin(txn, channel, channels)
+                                      }
+                                      .map { _ =>
+                                        logger.debug(s"""|consume: no data found,
                                  |storing <(patterns, continuation): ($patterns, $continuation)>
                                  |at <channels: $channels>""".stripMargin.replace('\n', ' '))
-                Right(None)
-              case Right(Some(dataCandidates)) =>
-                consumeCommCounter.increment()
-
-                eventLog.update(COMM(consumeRef, dataCandidates.map(_.datum.source)) +: _)
+                                        Right(None)
+                                      }
+                                  case Right(Some(dataCandidates)) =>
```

Comment:
I see why you used the syntax you did, because the `dataCandidates` traverse is the only monadic function. IMHO, this case is easier to grok if it's wrapped in a `for` comprehension.

Reply:
addressed in the "refactor" and "enable span" for rspace commit